Mozillians_Repositioning.md

```diff
@@ -0,0 +1,233 @@
+- new front-end renders profile view
+
+##### Editing a Profile
+
+- user is on the edit profile view
+- user clicks on ✎ in a section of their profile
+- hits Django
+- gets view with new front-end
+- new front-end requests edit profile section data from API (Django)
+- new front-end renders edit profile section view
+- user changes value
+- user clicks on save button
+- changed fields get sent to API
+- resulting partial profile update gets sent to [profile publisher](#profile-publisher)
+- on success user gets redirected to edit profile view
+
+##### Searching
+
+- user enters something in the search bar on any page
+- while typing the new front-end queries the suggestion API for suggestions
+- Django queries search service according to users permissions for suggestions
+- front-end displays suggestions
+- user hits enter / clicks on search / clicks on a suggested search
+- font-end sends search query to search API
+- Django queries search service according to users permissions
+- new-front end renders search results and updates the search bar to reflect the proper search syntax
+
+##### …
+
+
+### Search Service
+
+The search service will be built on top of elastic search. It should provide the following APIs:
+
+- query for an individual profile by username / user id and return the according profile data
+- a search that exposes most of ES syntax that returns results in one of two formats:
+  - list of usernames / user ids
+  - list of minimal profile data (everything needed to render a card)
+- suggestion / auto-complete for search
+
+The search service will only be accessible by Mozillians, profile publisher and integrity checker.
+Ideally we store raw profile v2 data and just add some prefixed properties if necessary.
+
+#### Profile Updates
+
+The search service will have an endpoint to receive profile updates. This endpoint will insert or
+update the profile it received.
+
+#### Privacy Levels
+
+The search service will be responsible of filtering fields according to privacy settings. We have
+two proposals on how we can achieve this.
+
+1. We use one index per privacy level. This enables simple search but we need to do the filtering on
+   every insertion of data.
+2. We use a single index and include privacy filters in the search query and do filter fields in the
+   result set. This enables simple insertion but requires more work during search time.
+
+
+### Profile Publisher
+
+The profile publisher is the central broker for dealing with profile updates. Its main
+responsibilities are:
+
+- receive user created updates from Mozillians and send them to CIS
+  - the request/respond cycle will wait until the profile was successfully updated in CIS or an
+    error/timeout occurred
+- receive update events from CIS and send profile updates to [search service](#search-service)
+  and [orgchart service](#orgchart-service)
+
+The profile publisher will be implemented in an async mindset. Since it's mainly IO bound this
```

Comment:
async with expected real-time changes right?
ie, we don't want a user to change their profile, and immediately go to an RP but the change isnt there yet because it was only 30s and the process takes 5min for ex.

Reply:
Yes, this is very much async as in pythons `asyncio`.